src/proxy/outbound.rs

```diff
@@ -369,28 +379,27 @@ impl OutboundConnection {
 
         let us = us.unwrap();
         // For case upstream server has enabled waypoint
-        if !us.workload.waypoint_addresses.is_empty() {
-            let waypoint_address = us.workload.choose_waypoint_address().unwrap();
-            // Even in this case, we are picking a single upstream pod and deciding if it has a remote proxy.
-            // Typically this is all or nothing, but if not we should probably send to remote proxy if *any* upstream has one.
-            let waypoint_workload = match self.pi.workloads.fetch_workload(&waypoint_address).await
-            {
-                Some(wl) => wl,
-                None => return Err(Error::UnknownWaypoint(waypoint_address, downstream)),
-            };
-            return Ok(Request {
-                // Always use HBONE here
-                protocol: Protocol::HBONE,
-                source: source_workload,
-                // Use the original VIP, not translated
-                destination: target,
-                destination_workload: Some(us.workload),
-                expected_identity: Some(waypoint_workload.identity()),
-                gateway: SocketAddr::from((waypoint_address, 15008)),
-                // Let the client remote know we are on the inbound path.
-                direction: Direction::Inbound,
-                request_type: RequestType::ToServerWaypoint,
-            });
+        match self.pi.workloads.find_waypoint(us.workload.clone()).await {
+            Ok(None) => {} // do nothing
+            Ok(Some(waypoint_us)) => {
+                let waypoint_workload = waypoint_us.workload;
+                let wp_socket_addr =
+                    SocketAddr::new(waypoint_workload.workload_ip, waypoint_us.port);
+                return Ok(Request {
+                    // Always use HBONE here
+                    protocol: Protocol::HBONE,
+                    source: source_workload,
+                    // Use the original VIP, not translated
+                    destination: target,
+                    destination_workload: Some(us.workload),
+                    expected_identity: Some(waypoint_workload.identity()),
+                    gateway: wp_socket_addr,
+                    // Let the client remote know we are on the inbound path.
+                    direction: Direction::Inbound,
```

Comment:
This looks very confusing to me, @howardjohn Can you elaborate it

Reply:
this is unchanged from before. `Direction::Inbound` is useful for logging, and the rest of this struct is unchanged in shape in an attempt to keep the PR size here as small as possible; although I agree we could probably refactor this into a more intuitive representation for an outgoing request

Reply:
The original code was copy+pasted from a prototype written in Go that had a different architecture :-) definitely can improve